lib/benchee/benchmark/runner.ex

```diff
@@ -218,32 +218,54 @@ defmodule Benchee.Benchmark.Runner do
   defp measure_iteration(
          scenario = %Scenario{function: function},
          scenario_context = %ScenarioContext{
-           num_iterations: 1
+           num_iterations: 1,
+           config: %{measure_memory: measure_memory}
          }
        ) do
     new_input = run_before_each(scenario, scenario_context)
-    {:memory, memory_usage_before} = :erlang.process_info(self(), :memory)
-    {microseconds, return_value} = :timer.tc(main_function(function, new_input))
-    {:memory, memory_usage_after} = :erlang.process_info(self(), :memory)
+    function = main_function(function, new_input)
+
+    {otp_version, _} =
+      :otp_release
+      |> :erlang.system_info()
+      |> to_string
+      |> Integer.parse()
+
+    {microseconds, memory_usage, return_value} =
+      measure_time_and_memory(function, measure_memory && otp_version > 18)
+
     run_after_each(return_value, scenario, scenario_context)
-    {microseconds, memory_usage_after - memory_usage_before}
+    {microseconds, memory_usage}
   end
 
   defp measure_iteration(
          scenario,
          scenario_context = %ScenarioContext{
-           num_iterations: iterations
+           num_iterations: iterations,
+           config: %{measure_memory: measure_memory}
          }
        )
        when iterations > 1 do
     # When we have more than one iteration, then the repetition and calling
     # of hooks is already included in the function, for reference/reasoning see
     # `build_benchmarking_function/2`
     function = build_benchmarking_function(scenario, scenario_context)
-    {:memory, memory_usage_before} = :erlang.process_info(self(), :memory)
-    {microseconds, _return_value} = :timer.tc(function)
-    {:memory, memory_usage_after} = :erlang.process_info(self(), :memory)
-    {microseconds, memory_usage_after - memory_usage_before}
+
+    {microseconds, memory_usage, _return_value} =
+      measure_time_and_memory(function, measure_memory)
+
+    {microseconds, memory_usage}
+  end
+
+  defp measure_time_and_memory(function, true) do
+    {microseconds, return_value} = :timer.tc(function)
+    {_, memory_usage} = MemoryMeasure.apply(function)
+    {microseconds, memory_usage, return_value}
```

Comment:
Does this mean the tests with memory measurements will run twice as long? If so, is it included in the time estimate?

What if instead we had a separate configuration like `memory_measure_time` or something like that? This would mean the two things - measuring memory and ips would be completely separate. I think it could make sense, since they are measured separate anyway. Additionally, because memory measurements are so consistent, it might make sense to default it to just a very small number of runs.

Reply:
At the moment, the run time remains unchanged. So if you're measuring memory usage, you'll be getting a significantly smaller sample size.

I think the UX for this feature is going to need some time to settle. There's going to be a balance between what is intuitive vs. what we just have to document and hope that folks actually read that we'll need to find, and I'm not sure we're quite there yet.

For example, I could make an argument that we could take just one memory usage measurement and document that choice, making it clear that if for some reason there is randomness in your benchmark, that won't be measured. Or, we could separate them entirely - that would work too!

I kind of want to release this as it is first and see how folks use it, and once we get a little more user feedback then that will help inform how things work later on.